### PR TITLE
[ET-VK] Lower eligible conv1d as conv2d over a singleton height dim

### LINKED ISSUE #22329: [ET-VK] conv1d runs at ~3% of the GPU's matmul throughput; routing it through the conv2d im2col path is 25x faster
### 🐛 Describe the bug

`conv1d.glsl` runs at roughly 3 percent of the throughput the same GPU reaches on a matmul with the same MAC count. It is the single reason the Whisper-tiny encoder loses to XNNPACK on Adreno.

Re-expressing the two `nn.Conv1d` layers of the Whisper-tiny encoder as `nn.Conv2d` with a singleton height dim, which routes them to the existing `conv2d_im2col` + `conv2d_gemm` path, is **25x faster** on the conv frontend and turns the whole encoder from 0.79x XNNPACK into 1.40x.

### Measurements

Samsung Galaxy S26 Ultra, Snapdragon SM8850 (Adreno 840), ExecuTorch at c27baa8031, fp32.

Whisper-tiny encoder end to end, interleaved A/B with the order reversed each round, warm-up iteration discarded, 60 timed iterations per arm:

| arm | median | vs XNNPACK |
| --- | --- | --- |
| Vulkan | 163.3 ms | 0.79x |
| XNNPACK | 129.7 ms | 1.00x |
| Vulkan, conv1d re-expressed as conv2d with H=1 | 92.4 ms | **1.40x** |

Per-dispatch GPU timestamps from the shader query pool, one encoder execution (153.4 ms of GPU time against 154 ms of wall clock, so this is entirely GPU bound):

| kernel | dispatches | ms | share |
| --- | --- | --- | --- |
| **conv1d_float** | **2** | **65.7** | **42.8%** |
| matmul_vec_texture3d_float | 8 | 22.0 | 14.3% |
| linear_vec_bias_texture3d_texture2d_float | 20 | 21.6 | 14.1% |
| softmax_texture3d_float | 4 | 16.8 | 11.0% |
| view_texture_float | 37 | 10.4 | 6.8% |
| clone_float | 32 | 9.0 | 5.9% |

Two dispatches, 43 percent of the runtime. Converting those to useful arithmetic:

| dispatch | shape | useful throughput |
| --- | --- | --- |
| `conv1d_float` gwg [3000, 384, 1] | 80 -> 384, k=3, L=3000 | 31.3 GFLOP/s |
| `conv1d_float` gwg [1500, 384, 1] | 384 -> 384, k=3, s=2, L=1500 | 27.6 GFLOP/s |
| `linear_vec_bias` gwg [384, 375, 1] | M=1500, N=1536, K=384 | 921 GFLOP/s |
| `linear_vec_bias` gwg [96, 375, 1] | M=1500, N=384, K=1536 | 900 GFLOP/s |

Same GPU, same run, seconds apart, and the two linear dispatches have an identical MAC count to each other. conv1d is 30x off what the machine demonstrably does.

### Why

Two things in the conv1d path:

1. `conv1d.glsl` computes one output element per invocation with no tiling, no shared memory and no register blocking. Every tap re-fetches both a weight texel and an input texel, so nothing is reused across invocations. The 384 -> 384 layer issues 1152 iterations x 2 texture fetches per invocation across 576k invocations, about 1.3 G texture fetches for 664 M MACs.

2. `conv1d_gwg` sets `z = div_up_4(batch)`, so the texel's four lanes span the batch dimension. At batch 1, which is every audio and sequence model I am aware of, three of every four lanes are dead. The `nchw_to_image` dispatch for a `[1, 80, 3000]` input is `[3000, 80, 1]`, a 4x over-allocated texture, against `[3000, 1, 20]` for the equivalent `[1, 80, 1, 3000]`.

### The conv2d path already solves this

`should_use_conv2d_im2col()` routes SlidingWindow conv2d to `conv2d_im2col.glsl` + `conv2d_gemm.glsl` when `groups == 1 && dilation == 1 && (device_is_mali() || c_out >= 128)`. conv1d has no equivalent and always dispatches the naive shader.

Isolated test, just the Whisper conv frontend (two convs plus gelu), same weights in both, both verified against the CPU reference at cosine 1.00000000 and max abs diff 4.9e-04:

| form | GPU time | dispatches |
| --- | --- | --- |
| `Conv1d` | 66.5 ms | conv1d 17.7 + conv1d 47.7 |
| `Conv2d`, H=1 | **2.64 ms** | im2col 0.06 + gemm 0.64 + im2col 0.19 + gemm 1.47 |

25x on the frontend, 27.6x on the conv dispatches alone.

Full encoder with the conv layers wrapped so they lower as conv2d: one delegate blob, cosine 0.99999684 against the CPU reference, and the 92.4 ms in the first table.

### Suggested fix

Rewrite `aten.convolution` with a 3-D input into `unsqueeze -> conv2d -> squeeze` before partitioning, so the existing conv2d machinery (im2col selection, layout tagging, weight prepacking) applies unchanged. I will send a PR.

Note that Whisper on Adreno additionally needs #22327 to be correct at all; this issue is only about speed.

### Versions

ExecuTorch at c27baa8031. Device: Samsung Galaxy S26 Ultra, Snapdragon SM8850, Adreno 840, Android 16.

cc @SS-JIA @manuelcandales @digantdesai @cbilgin

### PR DESCRIPTION
**Pull request by msluszniak**
Fixes #22329.

`conv1d.glsl` computes one output element per invocation with no tiling, no shared memory and no register blocking, and every tap re-fetches both a weight texel and an input texel, so nothing is reused between invocations. On top of that `conv1d_gwg` sets `z = div_up_4(batch)`, so the texel's four lanes span the batch dimension and three of every four are dead at batch 1.

Measured inside a single Whisper-tiny encoder execution on an Adreno 840, using the shader query pool:

| dispatch | useful throughput |
| --- | --- |
| `conv1d_float`, 80 -> 384, k=3, L=3000 | 31.3 GFLOP/s |
| `conv1d_float`, 384 -> 384, k=3, s=2, L=1500 | 27.6 GFLOP/s |
| `linear_vec_bias`, M=1500, N=1536, K=384 | 921 GFLOP/s |
| `linear_vec_bias`, M=1500, N=384, K=1536 | 900 GFLOP/s |

Same GPU, same run. Those two conv1d dispatches were 43% of the encoder's runtime.

`conv2d_im2col.glsl` + `conv2d_gemm.glsl` already solve this for 2-D convolutions. This pass rewrites 1-D convolutions into a 2-D convolution over a singleton height dim so the existing conv2d machinery (im2col selection, memory layout tagging, weight prepacking) handles them with no runtime changes.

The weight is reshaped in place through `set_param_tensor` rather than with a `view_copy` node, because `add_conv2d_node` reads it as a constant `TensorRef` and a runtime view would hand it a `TENSOR` instead.

### Eligibility

Only convolutions that are certain to reach the im2col path are rewritten, so nothing silently moves onto a shader that has not been compared against conv1d:

- `groups == 1` and unit dilation, matching `should_use_conv2d_im2col`
- `out_channels >= kIm2colMinCOut` (128), so the im2col path is selected on every vendor rather than only on Mali
- `kernel_size > 1`, leaving pointwise convs on `conv1d_pw`
- batch 1, since conv2d throws on batched input and a rewrite would turn a working conv1d into a hard error
- a constant weight, so it can be reshaped in place

### Results

Whisper-tiny encoder, Samsung Galaxy S26 Ultra (Adreno 840), fp32, interleaved A/B with the order reversed each round, warm-up iteration discarded, 60 timed iterations per arm. Both Vulkan arms include #22328, without which the encoder is not deterministic on this GPU at all.

| arm | median | vs XNNPACK |
| --- | --- | --- |
| Vulkan before | 153.2 ms | 0.84x |
| XNNPACK | 128.0 ms | 1.00x |
| Vulkan after | **88.8 ms** | **1.44x** |

Cosine against the CPU reference is 0.99999684 after the change.

Isolated, just the encoder's two convs plus gelu, GPU time from the query pool:

| | before | after |
| --- | --- | --- |
| conv dispatches | conv1d 17.7 + 47.7 ms | im2col 0.05 + gemm 0.64 + im2col 0.19 + gemm 1.47 ms |
| total | 66.5 ms | 3.28 ms |

20x on the frontend, and the output is unchanged at cosine 1.00000000, max abs diff 4.9e-04 against the CPU reference in both.

### Routing and correctness matrix

Eight conv1d configurations exported and run on device, with the dispatched shader read back from the query pool:

| case | shader after the pass | rewritten | cosine | max abs diff |
| --- | --- | --- | --- | --- |
| 80 -> 384, k=3 | `conv2d_im2col` + `conv2d_gemm` | yes | 1.00000000 | 2.0e-06 |
| 384 -> 384, k=3, s=2 | `conv2d_im2col` + `conv2d_gemm` | yes | 1.00000000 | 4.1e-06 |
| 128 -> 256, k=5, p=0 | `conv2d_im2col` + `conv2d_gemm` | yes | 1.00000012 | 3.1e-06 |
| 64 -> 64, k=3 | `conv1d_float` | no | 1.00000000 | 8.3e-07 |
| 256 -> 256, k=1 | `conv1d_pw_bias_texture` | no | 1.00000000 | 9.5e-07 |
| 256 -> 256, k=3, depthwise | `conv1d_dw_bias_texture` | no | 1.00000012 | 4.8e-07 |
| 128 -> 256, k=3, dilation 2 | `conv1d_float` | no | 1.00000000 | 1.2e-06 |
| 256 -> 256, k=3, groups 4 | `conv1d_float` | no | 1.00000000 | 9.5e-07 |

Every ineligible case keeps its original shader and every case matches the CPU reference.

`test_vulkan_backend_conv1d_as_conv2d` covers the rewritten path in CI; the two existing conv1d tests are grouped and depthwise, so they only exercise the untouched path.

---

### Depends on #22328

Moving the conv frontend onto the im2col + GEMM path exposes a pre-existing Adreno defect in `conv2d_gemm`, which dispatches with the same near-square local work group as `linear`. Distinct outputs over 60 executions on the Adreno 840, with `Linear.cpp` already fixed:

| | distinct / 60 |
| --- | --- |
| Whisper encoder, conv1d frontend | 1 |
| Whisper encoder, this PR | 9 |
| Whisper conv frontend alone, this PR | 13 |

This is not caused by the rewrite. `conv2d_gemm` is already reached today by any conv2d with `out_channels >= 128`, and it is nondeterministic there for the same reason. #22328 now fixes the shared `pick_xy_square_lwg` helper rather than one caller, and with it every model above is 1 / 60.

The 88.8 ms figure above was measured with only `Linear.cpp` fixed. With #22328 applied in full the encoder is 91.4 ms, still 1.34x XNNPACK, and correct on every run.

### Why rewrite rather than write a tiled conv1d shader

The 384 -> 384 dispatch moves 21.2 GB of texture traffic for 8.7 MB of compulsory traffic, so the shader is texture bound with essentially no reuse, and the batch-in-lanes waste is a 4x term next to that. A native tiled conv1d would fix the reuse the same way the GEMM does, and would additionally save the im2col materialisation, which is 0.25 ms of the 2.64 ms frontend above, roughly 10%. This pass gets the rest of it by reusing kernels that are already written, tuned and tested, and leaves a dedicated conv1d kernel available later as an optimisation rather than a prerequisite.

cc @SS-JIA @manuelcandales @digantdesai @cbilgin

### CODE REVIEW

File: backends/vulkan/_passes/conv1d_as_conv2d.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            return False
+        # The weight has to be a constant so it can be reshaped in place.
+        if not is_param_node(self._exported_program, weight_node):
+            return False
+        transposed, groups = node.args[6], node.args[8]
+        if transposed or groups != 1:
+            return False
+        # conv2d does not support batched input at all, so a rewrite there
+        # would turn a working conv1d into a throw.
+        if in_shape[0] != 1:
```
Comment (SS-JIA) [inline backends/vulkan/_passes/conv1d_as_conv2d.py]: [High] For a symbolic batch dimension, this comparison can specialize from the traced value `1` without proving every allowed runtime value is 1. The rewritten conv2d path rejects batched input, so a valid larger runtime batch can fail. Require a shape-range proof with lower and upper bounds both equal to 1; otherwise skip the rewrite.

_AI reviewer note: this comment was generated by an AI reviewer._

File: backends/vulkan/_passes/conv1d_as_conv2d.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        in_4d_shape = [in_shape[0], in_shape[1], 1, in_shape[2]]
+        w_4d_shape = [w_shape[0], w_shape[1], 1, w_shape[2]]
+        out_4d_shape = [out_shape[0], out_shape[1], 1, out_shape[2]]
+
+        # The weight tensor is contiguous, so inserting a singleton dim is a
+        # pure metadata change. A weight shared by several convs is only
+        # reshaped once; the rank check in _eligible() skips it afterwards.
+        weight = get_param_tensor(self._exported_program, weight_node)
+        assert weight is not None
+        set_param_tensor(
```
Comment (SS-JIA) [inline backends/vulkan/_passes/conv1d_as_conv2d.py:119]: [High] This mutates the shared parameter before validating all consumers. If one eligible conv and one ineligible op share the weight, the first rewrite changes the weight to rank 4 and leaves the later consumer using the wrong shape. Clone per rewritten consumer or require all consumers to be safely rewritten before mutating shared state.

_AI reviewer note: this comment was generated by an AI reviewer._

File: backends/vulkan/test/test_vulkan_delegate.py [hunk 3]
```diff
@@ -1065,6 +1065,33 @@ def forward(self, x):
...
+                    bias=True,
+                )
+
+            def forward(self, x):
+                return self.conv(x)
+
+        conv1d_module = Conv1dModule()
+        sample_inputs = (torch.randn(size=(1, 32, 64), dtype=torch.float32),)
+
+        self.lower_module_and_test_output(
```
Comment (SS-JIA) [inline backends/vulkan/test/test_vulkan_delegate.py:1119]: [Medium] Output equality does not prove `Conv1dAsConv2dPass` ran; the unchanged conv1d path can satisfy this test. Assert the transformed graph contains the intended 4-D convolution/views and add shared-weight plus symbolic-batch rejection cases.

_AI reviewer note: this comment was generated by an AI reviewer._

**Review by SS-JIA — COMMENTED** [review]
AI-assisted review. Inline findings below.

File: backends/vulkan/_passes/conv1d_as_conv2d.py (hunk 1 above)
Comment (msluszniak) [reply backends/vulkan/_passes/conv1d_as_conv2d.py]: Fixed: the pass now bails unless the batch dim is a concrete 1, so a symbolic batch stays on the conv1d path. Covered by `test_conv1d_as_conv2d_rejects_symbolic_batch`.

File: backends/vulkan/_passes/conv1d_as_conv2d.py (hunk 2 above)
Comment (msluszniak) [reply backends/vulkan/_passes/conv1d_as_conv2d.py:119]: Fixed: the rewrite is skipped when the weight has more than one user, so a shared parameter is never reshaped out from under another consumer. Covered by `test_conv1d_as_conv2d_rejects_shared_weight`.

File: backends/vulkan/test/test_vulkan_delegate.py (hunk 3 above)
Comment (msluszniak) [reply backends/vulkan/test/test_vulkan_delegate.py:1119]: Added graph-level tests in `test_vulkan_passes.py`: `test_conv1d_as_conv2d_rewrites_eligible_conv` asserts the convolution input is rank 4 and two `view_copy` nodes were inserted, plus the shared-weight and symbolic-batch rejection cases.

**Review by SS-JIA — APPROVED** [review]

**Comment by SS-JIA** [comment]
@msluszniak looks good, but lintrunner is failing

**Comment by msluszniak** [comment]
Pushed a ufmt fix for `conv1d_as_conv2d.py` and `test_vulkan_passes.py`; lintrunner is clean locally.
